Key rotation

max-b opened the pull request:
This uses the preshared password to generate an initial key. That
preshared key is then used as a derivation key for key rotations.

This does not actually provide forward secrecy as there's a
deterministic relationship between the initial preshared key and
subsequent encryption keys. Instead, it is an iterative step towards
implementing an olm/signal ratchet forward secrecy mechanism because it
exercises dynamic key selection.

max-b committed:
Cleanup existing e2ee code

Use proper logger, update documentation, fix types.

max-b committed:
Add noImplicitAny back to tsconfig

max-b committed:
Implement key rotation

This uses the preshared password to generate an initial key. That
preshared key is then used as a derivation key for key rotations.

This does not actually provide forward secrecy as there's a
deterministic relationship between the initial preshared key and
subsequent encryption keys. Instead, it is an iterative step towards
implementing an olm/signal ratchet forward secrecy mechanism because it
exercises dynamic key selection.

max-b commented:
@barath I think the easiest way to review the change/inspect the current implementation is to just review this file:

https://github.com/Invisv-Privacy/client-sdk-js/blob/7eba1c9898a935d6ef4b55dcb58a8c63273c4758/src/worker/e2ee.ts

barath commented on src/room/track/RemoteTrack.ts:
```diff
@@ -27,15 +28,18 @@ export default abstract class RemoteTrack extends Track {
     this.decryptTrack();
+    setInterval(() => {
+      this.worker.postMessage({ operation: 'rotateKey' });
+    }, 20000);
```
What's the reason there's a different time here on initialization (and should it be a defined constant)?

max-b replied:
it should be a defined constant and the reason it's not is because there's a change that got left out of that commit :facepalm:  

thanks for the catch

max-b committed:
Only leave 1 byte unencrypted in vp8 header

It looks as though that's all that's necessary, though we can keep an
eye on SFU performance moving forwards.

max-b commented:
Gonna go ahead and merge here and then pull this into our client UI
